google_guest_agent/events/events.go

```diff
@@ -0,0 +1,221 @@
+//  Copyright 2023 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// Package events is a events processing layer.
+package events
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/events/metadata"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+)
+
+var (
+	// availableWatchers mapps all kown available event watchers.
+	availableWatchers = make(map[string]Watcher)
+)
+
+// Watcher defines the interface between the events manager and the actual
+// watcher implementation.
+type Watcher interface {
+	// ID returns the watcher id.
+	ID() string
+	// Run implements the actuall "listening" strategy and emits a event "signal".
+	// It must return:
+	//   - [bool] if the watcher should renew(run again).
+	//   - [string] the event id.
+	//   - [interface{}] a event context data pointer further describing the event(if needed).
+	Run(ctx context.Context) (bool, string, interface{})
+}
+
+// Manager defines the interface between events management layer and the
+// core guest agent implementation.
+type Manager struct {
+	watchers    map[string]Watcher
+	subscribers map[string][]*eventSubscriber
+}
+
+// Config offers a mechanism for the consumers to configure the Manager
+// behavior.
+type Config struct {
+	Watchers []string
+}
+
+// EventCb defines the callback interface between watchers and subscribers.
+type EventCb func(evType string, data interface{}, evData interface{})
+
+type eventSubscriber struct {
+	data interface{}
+	cb   EventCb
+}
+
+type eventNotif struct {
+	leaving bool
+	evType  string
+	data    interface{}
+}
+
+func init() {
+	err := initWatchers([]Watcher{
+		metadata.New(),
+	})
+	if err != nil {
+		logger.Errorf("Failed to initialize watchers: %+v", err)
+	}
+}
+
+// init initializes the known available event watchers.
+func initWatchers(watchers []Watcher) error {
+	for _, curr := range watchers {
+		// Ignore & log if we are accidentaly not properly setting the id.
```

Comment:
Is this comment accurate? it says ignore and log but we're returning error?

Reply:
Post split leftover, will get it fixed. Thank you.

Reply:
Done.